overlord/devicestate/handlers_install.go

```diff
@@ -1214,14 +1262,115 @@ func (m *DeviceManager) doInstallFinish(t *state.Task, _ *tomb.Tomb) error {
 	if err := t.Get("on-volumes", &onVolumes); err != nil {
 		return err
 	}
-	logger.Debugf("install-finish for %q on %v", systemLabel, onVolumes)
+	logger.Debugf("starting install-finish for %q on %v", systemLabel, onVolumes)
+
 	// TODO: use the seed to get gadget/kernel
 	// - install missing volumes structure content
 	// - install gadget assets
-	// - install kenrel
+	// - install kernel
 	// - make system bootable (include writing modeenv)
 
-	return fmt.Errorf("finish install step not implemented yet")
+	st.Unlock()
+	sys, snapInfos, snapSeeds, err := m.loadSystemAndEssentialSnaps(systemLabel,
+		[]snap.Type{snap.TypeKernel, snap.TypeBase, snap.TypeGadget})
+	st.Lock()
+	if err != nil {
+		return err
+	}
+	// Mount gadget and kernel
+	mntPtForType := make(map[snap.Type]string)
+	for _, seedSn := range []*seed.Snap{snapSeeds[snap.TypeGadget], snapSeeds[snap.TypeKernel]} {
+		mntPt, rest, err := mountSeedSnap(seedSn)
+		mntPtForType[seedSn.EssentialType] = mntPt
+		if err != nil {
+			return err
+		}
+		defer func() {
+			errRest := rest()
+			if errRest != nil {
+				err = errRest
+			}
+		}()
+	}
+
+	// unset revision means the snap will be local, we do something
+	// similar in the seedwriter
+	// XXX: is this the right place?
+	// TODO: also do this for gadget
+	kernelInfo := snapInfos[snap.TypeKernel]
+	if kernelInfo.Revision.Unset() {
+		kernelInfo.Revision = snap.R(-1)
+	}
+
+	// TODO useEncryption equals true case
+	useEncryption := false
+	installObserver, trustedInstallObserver, err := buildInstallObserver(sys.Model, mntPtForType[snap.TypeGadget], useEncryption)
+	if err != nil {
+		return err
+	}
+
```

Comment:
this needs to mount the partitions like install.Run would, as that is not a task for the installer as such